File: packages/python-sdk/strait/composition/_checkpoint_resume.py
"""Checkpoint resume helper for resumable runs."""

from __future__ import annotations

from typing import Any, Awaitable, Callable, TypeVar

from strait.authoring._run_context import RunContext

T = TypeVar("T")


async def with_checkpoint_resume(
    ctx: RunContext,
    last_checkpoint: dict[str, Any] | None,
    fn: Callable[[dict[str, Any], Callable[[dict[str, Any]], None]], Awaitable[T]],
    initial_state: dict[str, Any],
    checkpoint_interval: int = 1,
) -> T:
    current_state = last_checkpoint if last_checkpoint is not None else initial_state
    step_count = 0

    def update_state(new_state: dict[str, Any]) -> None:
        nonlocal current_state, step_count
        current_state = new_state
        step_count += 1
        if step_count % checkpoint_interval == 0 and ctx.checkpoint is not None:
            try:
                import asyncio
                coro = ctx.checkpoint(current_state)
                if asyncio.iscoroutine(coro):
                    asyncio.ensure_future(coro)
            except Exception:
                pass

    result = await fn(current_state, update_state)

    if ctx.checkpoint is not None:
        await ctx.checkpoint(current_state)

    return result

File: packages/typescript-sdk/src/composition/checkpoint-resume.ts
import type { RunContext } from "../authoring/job";

export type CheckpointResumeOptions<TState extends Record<string, unknown>> = {
  readonly initialState: TState;
  readonly checkpointInterval?: number;
};

export const withCheckpointResume = async <
  TState extends Record<string, unknown>,
  TOutput,
>(
  ctx: RunContext,
  lastCheckpoint: TState | undefined,
  fn: (
    state: TState,
    updateState: (newState: TState) => void
  ) => Promise<TOutput>,
  options: CheckpointResumeOptions<TState>
): Promise<TOutput> => {
  let currentState = lastCheckpoint ?? options.initialState;
  const interval = options.checkpointInterval ?? 1;
  let stepCount = 0;

  const updateState = (newState: TState) => {
    currentState = newState;
    stepCount++;
    if (stepCount % interval === 0) {
      ctx.checkpoint(currentState).catch(() => undefined);
    }
  };

  const result = await fn(currentState, updateState);

  // Final checkpoint
  await ctx.checkpoint(currentState);

  return result;
};
